### PR TITLE
Some work on Attack*

### PR DESCRIPTION
**Pull request by matija-hustic**
Some rewiring (not a full refactor) in the attacking mechanism to accommodate using different cursors and voices with different armaments, as well as combining beneficial (buffs) and damaging weapons for a single actor.

Eliminates the need for the medic to subclass `AttackFrontal` and `AutoHeal` or use `Heal`.

I'll type up the upgrade rule after review of other code.

Discussed to some extent in #8763. Prerequisite for #8563.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 1]
```diff
@@ -62,6 +62,19 @@ public class ArmamentInfo : UpgradableTraitInfo, Requires<AttackBaseInfo>
 		[Desc("Use multiple muzzle images if non-zero")]
 		public readonly int MuzzleSplitFacings = 0;
 
+		[Desc("Is it possible to use the armament when force-attacking?")]
+		public readonly bool ForceAttacks = true;
+
+		[Desc("Is the armament beneficial (should it be used on allies)?")]
+		public readonly bool Beneficial = false;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Armament.cs]: Why not name this `CanTargetAllies`?

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Armament.cs]: Force-attack ignores this tag. Keep in mind this is used for target acquisition to define it. Also, `CanTargetAllies: false` wouldn't suggest that the weapon is used on enemies.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Armament.cs]: No, it wouldn't, as those are different things. `Beneficial` doesn't sound very good to me too. Let's try to come up with an alternative.
`IsUsedOnAllies` ?  `AutoTargetsAllies` ?

Comment (matija-hustic) [inline OpenRA.Mods.Common/Traits/Armament.cs]: I was thinking along the lines of `MeantForAllies` or `CannotTargetEnemies`

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Armament.cs]: See https://github.com/Ares-Developers/Ares/blob/master/src/Ext/Techno/Hooks.cpp#L632 why any `*Allies` are wrong.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Armament.cs]: All `*Allies` can't be wrong since this weapon is going to only be used on allies.
So something like `IsUsedOnlyOnAllies` ( + `WhenAutoTargeting`).

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Armament.cs]: `IntendedForAllies` then. But really, any `Allies` tag would mean - with the long term plan to remove the fragile alliances - that there should be an `Enemies` counterpart which I'm definitely not a fan of.

Comment (matija-hustic) [inline OpenRA.Mods.Common/Traits/Armament.cs]: This is the only place where `Beneficial` is used

https://github.com/OpenRA/OpenRA/pull/8768/files#diff-26d2cbcfb01852edd1172678f8116784R244

Right now,
* if an armament has `Beneficial == false`,
  * it can be used to target enemies as usual
    * without force-attack,
    * and with it if additionally `ForceAttacks == true`
  * and allies
    * only using force-attack if additionally `ForceAttacks == true`
* if `Beneficial == true`,
  * only allies can be targeted with it
    * without force-attack,
    * and with it if additionally `ForceAttacks == true`
  * enemies cannot be targeted at all
    * neither with
    * nor without force-attack

**Comment by UberWaffe** [comment]
A suggestion:

"NotHarmful" instead of "Beneficial" ?
That would imply that a unit wouldn't care about getting hit by it.

A healing serum would be not harmful, while a targeting laser would be harmful (even though it doesn't directly cause harm, it still results in harm.)

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (UberWaffe) [inline OpenRA.Mods.Common/Traits/Armament.cs]: A suggestion:

"NotHarmful" instead of "Beneficial" ?
That would imply that a unit wouldn't care about getting hit by it, regardless of its diplomatic stance.

A healing serum would be not harmful, while a targeting laser would be harmful (even though it doesn't directly cause harm, it still results in harm.)

**Comment by abcdefg30** [comment]
RA medic and mechanic still work as well as the testcase. :+1: / :white_check_mark:

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Armament.cs]: So, what about this now? Leave as-is or change to a suggestion?

**Comment by Unit158** [comment]
Needs a rebase

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Armament.cs]: You can remove the confusion and make it more consistent with our other traits by defining:
```csharp
public readonly Stance Targets: Stance.Enemy;
public readonly Stance ForceTargets: Stance.Enemy | Stance.Neutral | Stance.Ally;
```

File: OpenRA.Mods.Common/Warheads/DamageWarhead.cs [hunk 2]
```diff
@@ -48,13 +48,27 @@ public int DamageVersus(ActorInfo victim)
...
 				return false;
 
-			return DamageVersus(victim) > 0;
+			if (Damage < 0 && victim.GetDamageState() == DamageState.Undamaged)
+				return false;
+
+			return DamageVersus(victim.Info) > 0;
+		}
+
+		// TODO: This can be removed after the legacy and redundant 0% = not targetable
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: This assumption has now been removed.

**Comment by pchote** [comment]
This is a good step towards a much-needed cleanup, so its a shame to see it sit idle.  Are you able to work on this any time soon @matija-hustic, or should somebody else take this over?

**Comment by obrakmann** [comment]
I think [these two lines](https://github.com/matija-hustic/OpenRA/blob/attackbase_refactor/OpenRA.Mods.Common/Traits/AutoTarget.cs#L131-L132) need to be changed to as well to fix #8481 and #9266. If `currentTarget` is not null, you'll want `range` to be the range of a valid weapon against it, not `attack.GetMaximumRange()`.

**Comment by matija-hustic** [comment]
Hey guys, I'm still in the middle of moving to a new city and I'm gonna need at least another week before I'm able to do anything. The time it's taking has surpassed even my own initial predictions :( It's probably going to take another week. If you're in a hurry to get my PRs off the queue, have someone else take over what ever they can.

**Comment by matija-hustic** [comment]
I've updated this. #8184 should no longer happen. I haven't checked #9266, though.

**Comment by matija-hustic** [comment]
I've made `ChooseArmamentsForTarget` ignore disabled armaments and now it seems #9266 is gone, too. At least I haven't been able to reproduce it.

**Comment by Phrohdoh** [comment]
Yes that would be one fix for that.

FWIW

I'd personally prefer a field like such on AttackBase:
```csharp
    EnabledArmaments { return Armaments.Where(a => !a.IsTraitDisabled); }
```

Or perhaps a `GetEnabledArmaments()` method to indicate work being done.

**Comment by matija-hustic** [comment]
@Phrohdoh, how about i put a `bool Enabled = true` in the `ChooseArmamentsForTarget`'s arguments?
Or do you need a separate function?

**Comment by Phrohdoh** [comment]
That ought to suffice though I think a more fitting name would be `onlyEnabled` or something to that effect.

**Comment by matija-hustic** [comment]
Done

**Comment by reaperrr** [comment]
I'm putting `Has Dependencies` on this so it doesn't get accidentally merged with testcase and without upgrade rules.

**Comment by reaperrr** [comment]
Also needs a rebase now.

**Comment by matija-hustic** [comment]
Rebased. Upgrade rule now in there. I'll leave the `TEST Medic` commit in there until the final review.

Otherwise, I've checked and #9266 should be gone, too, now, not only #8184.

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 3]
```diff
@@ -64,8 +67,16 @@ protected virtual Activity InnerTick(Actor self, AttackBase attack)
...
 				return NextActivity;
 
+			// Drop the target once none of the weapons are effective against it
+			var armaments = attack.ChooseArmamentsForTarget(Target, forceAttack);
+			if (!armaments.Any())
+				return NextActivity;
+
+			// Update ranges
+			minRange = armaments.MaxBy(a => a.Weapon.MinRange.LengthSquared).Weapon.MinRange;
+			maxRange = armaments.MinBy(a => a.Weapon.Range.LengthSquared).MaxRange();
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: You need to order by `a.MaxRange()`, which includes range modifiers, instead of `a.Weapon.Range`.

Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: `armaments.Min(a => a.MaxRange())` should work, with a similar simplification for `minRange`.

**Comment by matija-hustic** [comment]
Changed as commented

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 4]
```diff
@@ -65,6 +65,17 @@ public class ArmamentInfo : UpgradableTraitInfo, IRulesetLoaded, Requires<Attack
 		public WeaponInfo WeaponInfo { get; private set; }
 		public WDist ModifiedRange { get; private set; }
 
+		public readonly Stance Targets = Stance.Enemy;
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Armament.cs]: `Targets` could be confused with target types. Perhaps `TargetAllegiances` or possibly just `Allegiances`.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Armament.cs]: Our convention is to use "stance" when referring to stances/allegiances.

**Comment by atlimit8** [comment]
The "PR: Needs +2" was outdated anyways.

**Comment by matija-hustic** [comment]
Changed

**Comment by obrakmann** [comment]
Works, doesn't break anything afaics, and code looks ok, too :+1:

**Comment by atlimit8** [comment]
:+1: - Test commit needs to be remove before merging,

**Comment by reaperrr** [comment]
Also needs a rebase now.

**Comment by matija-hustic** [comment]
OK, so the test commit is out and the branch rebased

**Comment by penev92** [comment]
Can you (if possible) change the first commit's message to something more descriptive than `Some work on Attack*`? This is a lot of changes, so I assume there is a lot to put in there.

**Comment by matija-hustic** [comment]
Ah yes, I forgot that the commits should be more descriptive. Is it better now, @penev92 ?

**Comment by penev92** [comment]
I can't verify that what it says is correct because I haven't reviewed this, but it looks a lot better, thanks ;)

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/6dbb6a0ce3301ce0e346b2ba1881be09e264d4e2)

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 5]
```diff
@@ -2198,6 +2198,52 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
 						node.Key = "-WithSpriteRotorOverlay";
 				}
 
+				if (engineVersion < 20151005)
+				{
+					// Move certain properties from Parachutable to new WithParachute trait
+					// Add dependency traits to actors implementing Parachutable
+					// Make otherwise targetable parachuting actors untargetable
```
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: This comment seems to be unrelated copy & pasta from https://github.com/matija-hustic/OpenRA/blob/69696fc154ebdf596cbb77c438f4f091b552d2d5/OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs#L1569

Comment (matija-hustic) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: grrrrr

https://github.com/matija-hustic/OpenRA/commit/9927057c0f76058959141dcecf5816886398f83f

Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Thanks for the quick fix in https://github.com/OpenRA/OpenRA/pull/9581.
